### ISSUE TITLE
Discrepancy between pyright CLI and pylance extension

### ISSUE THREAD

**Issue by kaysond** [microsoft/pyright#5664]
**Describe the bug**
Consider the file:
```python
# pyright: strict
def main():
    return sys.argv


if __name__ == "__main__":
    import sys

    main()
```

If I run `pyright` on this file via CLI, I get no errors. VS Code, via the pylance extension, however complains about a lot of things because of the import in a conditional, among them:
```json
[{
	"resource": "test.py",
	"owner": "_generated_diagnostic_collection_name_#6",
	"code": {
		"value": "reportUnboundVariable",
		"target": {
			"$mid": 1,
			"path": "/microsoft/pyright/blob/main/docs/configuration.md",
			"scheme": "https",
			"authority": "github.com",
			"fragment": "reportUnboundVariable"
		}
	},
	"severity": 8,
	"message": "\"sys\" is unbound",
	"source": "Pylance",
	"startLineNumber": 3,
	"startColumn": 12,
	"endLineNumber": 3,
	"endColumn": 15
}]
```
I couldn't find a way to make the extension show me how it's using pyright, even with trace logs being on. I wondered if it's related to import resolution, but that would mean the CLI is resolving `sys` always, even if the conditional turns out to be false, and I don't think that's the intended behavior.

**VS Code extension or command-line**
VS Code Pylance Extension: v2023.8.10
Pyright version: 1.1.320

**Comment by erictraut** [comment]
This is due to a bug fix in 1.1.320. The latest production version of Pylance uses 1.1.319. If you switch to the prerelease version of Pylance (v2023.8.11), you'll see that the behavior is consistent with 1.1.320.

**Comment by kaysond** [comment]
> This is due to a bug fix in 1.1.320. The latest production version of Pylance uses 1.1.319. If you switch to the prerelease version of Pylance (v2023.8.11), you'll see that the behavior is consistent with 1.1.320.

@erictraut so to confirm - you're saying the intended behavior is that even if an import is in a conditional, pyright will resolve all of its uses in the file? Even though it could cause an error if the file is imported and the function is called without the import?

**Comment by erictraut** [comment]
The behavior that you see in pyright 1.1.320 is the intended behavior. A [bug](https://github.com/microsoft/pyright/issues/5618) was introduced in 1.1.319 that resulted in the unintended behavior. It was promptly fixed in 1.1.320.

Pyright generally doesn't report "potentially unbound" errors for symbols that are assigned in other execution scopes because a static type checker cannot know the order in which different scopes are executed. If pyright were to report these errors, it would result in many false positives for many legitimate use cases. In cases where pyright can determine (without the potential for false positives) that the symbol is unbound, it will report the error in that case. For more details, refer to [this issue](https://github.com/microsoft/pyright/issues/5548).
